File: server/jobs.py
"""Single-book subprocess queue with explicit cancellation and observable health."""
from __future__ import annotations

import contextlib
import fcntl
import os
import signal
import subprocess
import sys
import threading
import time


class BusyBook(RuntimeError):
    pass


@contextlib.contextmanager
def book_lease(root):
    """Exclude a separate pipeline process without trusting persisted status or PID files."""
    work = root / 'work'
    work.mkdir(exist_ok=True)
    with (work / 'run.lock').open('a') as handle:
        try:
            fcntl.flock(handle, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BlockingIOError as exc:
            raise BusyBook('这本书正在由另一个任务处理，请先停止该任务') from exc
        try:
            yield
        finally:
            fcntl.flock(handle, fcntl.LOCK_UN)


class Worker(threading.Thread):
    def __init__(self, books, app, read, write, enabled=True):
        super().__init__(daemon=True, name='yedu-book-worker')
        self.books, self.app, self.read, self.write = books, app, read, write
        self.enabled = enabled
        self.mode = os.environ.get('YEDU_WORKER_MODE', 'process')
        if self.mode not in ('process', 'thread'):
            raise ValueError('Invalid worker mode')
        self.wake = threading.Event()
        self.manual_queue = set()
        self.stopping = threading.Event()
        self.lock = threading.RLock()
        self.current = None
        self.process = None
        self.cancel_event = None
        self.current_done = threading.Event()
        self.current_done.set()
        self.last_error = None
        self.last_scan = 0

    def health(self):
        with self.lock:
            return {'enabled': self.enabled, 'alive': self.is_alive(), 'current': self.current,
                    'mode': self.mode,
                    'pid': self.process.pid if self.process and self.process.poll() is None else None,
                    'last_scan': self.last_scan, 'last_error': self.last_error}

    def set_auto(self, root, enabled):
        with self.lock:
            meta = dict(self.read(root / 'meta.json') or {})
            state = dict(self.read(root / 'status.json') or {})
            quality = state.get('quality') or {}
            retry_quality = (enabled and bool(quality.get('pending') or quality.get('state') == 'pending')
                             and state.get('state') not in ('running', 'finalizing', 'cancelling'))
            meta['auto'] = enabled
            if retry_quality:
                # This is an explicit user action. Startup never infers permission
                # to regenerate completed books from a pending-quality marker.
                meta['retry_quality'] = True
            elif not enabled:
                meta.pop('retry_quality', None)
            self.write(root / 'meta.json', meta)
            if enabled and (retry_quality or state.get('state') not in ('done', 'running', 'finalizing', 'cancelling')):
                state.update(state='queued', updated=time.time(), error=None)
                self.write(root / 'status.json', state)
            if not self.enabled:
                if enabled:
                    self.manual_queue.add(root.name)
                else:
                    self.manual_queue.discard(root.name)
            self.wake.set()

    def cancel(self, root, timeout=15, preserve_auto=False):
        if not preserve_auto:
            self.set_auto(root, False)
        thread_active = False
        with self.lock:
            process = self.process if self.current == root.name else None
            if self.mode == 'thread' and self.current == root.name and self.cancel_event is not None:
                thread_active = True
                self.cancel_event.set()
                state = dict(self.read(root / 'status.json') or {})
                state.update(state='cancelling', updated=time.time())
                self.write(root / 'status.json', state)
            if process and process.poll() is None:
                state = dict(self.read(root / 'status.json') or {})
                state.update(state='cancelling', updated=time.time())
                self.write(root / 'status.json', state)
                try:
                    os.killpg(process.pid, signal.SIGTERM)
                except ProcessLookupError:
                    pass
        if process:
            try:
                process.wait(timeout=timeout)
            except subprocess.TimeoutExpired:
                try:
                    os.killpg(process.pid, signal.SIGKILL)
                except ProcessLookupError:
                    pass
                process.wait(timeout=5)
        if thread_active and not self.current_done.wait(timeout):
            # The current model call may still be finishing. Its book lease stays held,
            # and the worker will publish paused/queued when it reaches a boundary.
            return
        # External jobs have no verified ownership: do not kill them by a guessed PID.
        with self.lock, book_lease(root):
            state = dict(self.read(root / 'status.json') or {})
            if state.get('state') != 'done':
                state.update(state='queued' if preserve_auto else 'paused', updated=time.time(), error=None)
                self.write(root / 'status.json', state)

    def _one(self, root):
        with self.lock:
            if not (root / 'book.json').exists():
                return
            meta = self.read(root / 'meta.json') or {}
            state = self.read(root / 'status.json') or {}
            retry_quality = bool(meta.get('retry_quality'))
            if not meta.get('auto') or (state.get('state') == 'done' and not retry_quality):
                return
            if state.get('state') == 'error' and time.time() - state.get('updated', 0) < 1800:
                return
            try:
                with book_lease(root):
                    pass
            except BusyBook:
                return
            retry_journal = root / 'work' / 'quality-retry.json'
            retry_before = self.read(retry_journal) if retry_quality else None
            self.current = root.name
            self.current_done.clear()
            try:
                if self.mode == 'thread':
                    self.cancel_event = threading.Event()
                    log = None
                    process = None
                else:
                    log = (root / 'work.log').open('a')
                    command = [sys.executable, '-u', '-m', 'pipeline.run', str(root)]
                    if retry_quality:
                        command.append('--retry-quality')
                    self.process = subprocess.Popen(
                        command, cwd=self.app,
                        stdout=log, stderr=subprocess.STDOUT, env=dict(os.environ), start_new_session=True)
                    process = self.process
            except Exception:
                self.current = None
                self.process = None
                self.cancel_event = None
                self.current_done.set()
                if 'log' in locals() and log is not None:
                    log.close()
                raise
        try:
            if self.mode == 'thread':
                from pipeline.run import AlreadyRunning, Cancelled, run_book
                try:
                    run_book(root, retry_quality=retry_quality,
                             model=os.environ.get('EXTRACT_MODEL', 'deepseek-flash+nothink'),
                             local_model=os.environ.get('LOCAL_MODEL', 'deepseek-flash+nothink'),
                             concurrency=int(os.environ.get('LOCAL_CONCURRENCY', '4')),
                             cancel_event=self.cancel_event)
                    code = 0
                except (Cancelled, AlreadyRunning):
                    code = 75
                except Exception as exc:
                    code = 1
                    self.last_error = {'book': root.name, 'message': str(exc)[:200], 'at': time.time()}
            else:
                code = process.wait()
            with self.lock:
                if not root.exists():
                    return
                state = dict(self.read(root / 'status.json') or {})
                meta = dict(self.read(root / 'meta.json') or {})
                auto = meta.get('auto')
                if retry_quality and code != 75:
                    retry_after = self.read(retry_journal) or {}
                    acknowledged = (retry_after.get('state') in ('rebuilding', 'complete')
                                    and (retry_after != retry_before
                                         or (retry_before or {}).get('state') == 'rebuilding'))
                    if acknowledged:
                        # Preserve the explicit request across a killed launch;
                        # only the child's durable rebuild journal acknowledges it.
                        meta.pop('retry_quality', None)
                        self.write(root / 'meta.json', meta)
                if auto and self.stopping.is_set() and code != 0:
                    state.update(state='queued', updated=time.time(), error=None)
                    self.write(root / 'status.json', state)
                elif not auto:
                    if state.get('state') != 'done':
                        state.update(state='paused', updated=time.time(), error=None)
                        self.write(root / 'status.json', state)
                elif code != 75 and (code != 0 or state.get('state') != 'done'):
                    state.update(state='error', updated=time.time(),
                                 error=state.get('error') or f'处理进程未完成（退出码 {code}）')
                    self.write(root / 'status.json', state)
                    self.last_error = {'book': root.name, 'message': state['error'], 'at': time.time()}
        finally:
            if log is not None:
                log.close()
            with self.lock:
                self.current = None
                self.process = None
                self.cancel_event = None
                self.current_done.set()

    def run(self):
        while not self.stopping.is_set():
            self.wake.wait(30)
            self.wake.clear()
            self.last_scan = time.time()
            try:
                if self.enabled:
                    roots = sorted((p.parent for p in self.books.glob('*/meta.json')
                                    if not p.parent.name.startswith('.')), key=lambda p: p.name)
                else:
                    with self.lock:
                        roots = [self.books / bid for bid in sorted(self.manual_queue)]
                        self.manual_queue.clear()
                for root in roots:
                    if self.stopping.is_set():
                        break
                    try:
                        self._one(root)
                    except Exception as exc:
                        self.last_error = {'book': root.name, 'message': str(exc)[:200], 'at': time.time()}
                        if root.exists():
                            try:
                                state = dict(self.read(root / 'status.json') or {})
                                state.update(state='error', error='后台处理失败，请查看服务日志', updated=time.time())
                                self.write(root / 'status.json', state)
                            except Exception:
                                pass
            except Exception as exc:
                self.last_error = {'message': str(exc)[:200], 'at': time.time()}

    def stop(self):
        self.stopping.set()
        if self.cancel_event is not None:
            self.cancel_event.set()
        self.wake.set()

File: core/lib/src/server/jobs.dart
/// Serialized book worker (`server/jobs.py`) for the native Dart application.
///
/// Existing meta/status and quality-retry files remain authoritative. The
/// additional worker receipt records ownership transitions, never model secrets.
library;

import 'dart:async';
import 'dart:collection';
import 'dart:convert';
import 'dart:io';

import '../async_util.dart';
import '../env.dart';
import '../errors.dart';
import '../pipeline/jev.dart' as jev;
import '../pipeline/run.dart' as pipeline;
import '../pipeline/run_lease.dart';
import '../py/py_compat.dart';
import 'storage.dart';

export '../pipeline/run_lease.dart'
    show RunCancellation, Cancelled, AlreadyRunning, RunLease, isBookRunning;

typedef Json = Map<String, Object?>;
typedef WorkerRun =
    Future<void> Function(
      Directory root, {
      required RunCancellation cancellation,
      required bool retryQuality,
      required String model,
      required String localModel,
      required int concurrency,
    });

final class BusyBook extends RuntimeError {
  const BusyBook() : super('这本书正在由另一个任务处理，请先停止该任务');
}

/// Execute while owning the same kernel lease used by Python's book_lease.
Future<T> bookLease<T>(Directory root, FutureOr<T> Function() action) async {
  final RunLease lease;
  try {
    lease = await RunLease.acquire(root);
  } on AlreadyRunning {
    throw const BusyBook();
  }
  try {
    return await action();
  } finally {
    lease.release();
  }
}

/// Resolved immediately before each job; UI settings stay outside this module.
class WorkerSettings {
  const WorkerSettings({
    this.model = 'deepseek-flash+nothink',
    this.localModel = 'deepseek-flash+nothink',
    this.concurrency = 4,
  });

  factory WorkerSettings.environment() => WorkerSettings(
    model: environ['EXTRACT_MODEL'] ?? 'deepseek-flash+nothink',
    localModel: environ['LOCAL_MODEL'] ?? 'deepseek-flash+nothink',
    concurrency: int.parse(environ['LOCAL_CONCURRENCY'] ?? '4'),
  );

  final String model;
  final String localModel;
  final int concurrency;
}

Future<void> _run(
  Directory root, {
  required RunCancellation cancellation,
  required bool retryQuality,
  required String model,
  required String localModel,
  required int concurrency,
}) async {
  await pipeline.runBook(
    root,
    cancellation: cancellation,
    retryQuality: retryQuality,
    model: model,
    localModel: localModel,
    concurrency: concurrency,
  );
}

Json? _readJson(File file) {
  if (!file.existsSync()) return null;
  final Object? value = jsonDecode(file.readAsStringSync());
  if (value is! Json) throw const ValueError('任务状态文件格式无效');
  return value;
}

double _now() => DateTime.now().millisecondsSinceEpoch / 1000;
bool _truth(Object? v) =>
    v != null &&
    v != false &&
    v != 0 &&
    v != '' &&
    !(v is Iterable<Object?> && v.isEmpty) &&
    !(v is Map<Object?, Object?> && v.isEmpty);
Json _object(Object? v) => v as Json? ?? <String, Object?>{};
String _name(Directory root) =>
    root.uri.pathSegments.where((String s) => s.isNotEmpty).last;
String _short(Object error) => PyCompat.slice(error.toString(), 0, 200);
bool _same(Object? a, Object? b) {
  if (a is Map<String, Object?> && b is Map<String, Object?>) {
    return a.length == b.length &&
        a.keys.every((String k) => b.containsKey(k) && _same(a[k], b[k]));
  }
  if (a is List<Object?> && b is List<Object?>) {
    return a.length == b.length &&
        Iterable<int>.generate(a.length).every((int i) => _same(a[i], b[i]));
  }
  return a == b;
}

/// Application-owned worker. Construct and use in one isolate; start() never
/// enables automatic paid requests. The default mode only runs explicit actions.
class Worker {
  Worker(
    this.books, {
    this.enabled = false,
    WorkerRun? run,
    WorkerSettings Function()? settings,
    this.onChange,
    Json? Function(File)? read,
    void Function(File, Json)? write,
    Future<bool> Function(Directory)? probe,
    double Function()? clock,
    this.scanInterval = const Duration(seconds: 30),
  }) : _runBook = run ?? _run,
       _settings = settings ?? WorkerSettings.environment,
       _read = read ?? _readJson,
       _write = write ?? writeJson,
       _probe = probe ?? isBookRunning,
       _clock = clock ?? _now;

  final Directory books;
  final bool enabled;
  final void Function(Json)? onChange;
  final Duration scanInterval;
  final WorkerRun _runBook;
  final WorkerSettings Function() _settings;
  final Json? Function(File) _read;
  final void Function(File, Json) _write;
  final Future<bool> Function(Directory) _probe;
  final double Function() _clock;
  final LinkedHashSet<String> _queue = LinkedHashSet<String>();
  final Semaphore _processGate = Semaphore(1);
  Future<void>? _initializing;
  Future<void>? _draining;
  Completer<void>? _currentDone;
  RunCancellation? _cancellation;
  Timer? _timer;
  bool _alive = false;
  bool _stopping = false;
  bool _scanAgain = false;
  int _sequence = 0;
  String? _current;
  Json? _lastError;
  double _lastScan = 0;

  Json health() => <String, Object?>{
    'enabled': enabled,
    'alive': _alive,
    'current': _current,
    'mode': 'dart',
    'pid': null,
    'last_scan': _lastScan,
    'last_error': _lastError == null ? null : <String, Object?>{..._lastError!},
    'queued': _queue.toList()..sort(),
    'stopping': _stopping,
  };

  void _notify() {
    // A disconnected UI observer cannot terminate or relaunch the book task.
    try {
      onChange?.call(health());
    } on Object {
      /* Observer is not the worker. */
    }
  }

  File _file(Directory root, String path) => File('${root.path}/$path');
  Json _json(Directory root, String path) => <String, Object?>{
    ...?_read(_file(root, path)),
  };
  void _save(Directory root, String path, Json value) =>
      _write(_file(root, path), value);

  Future<void> _withLease(Directory root, void Function() action) async {
    await bookLease<void>(root, action);
  }

  void _checkRoot(Directory root) {
    if (!root.existsSync() ||
        !books.existsSync() ||
        root.parent.resolveSymbolicLinksSync() !=
            books.resolveSymbolicLinksSync() ||
        _name(root).startsWith('.') ||
        FileSystemEntity.typeSync(root.path, followLinks: false) !=
            FileSystemEntityType.directory) {
      throw const ValueError('书籍目录无效');
    }
  }

  List<Directory> _roots() {
    if (!books.existsSync()) return <Directory>[];
    return books
        .listSync(followLinks: false)
        .whereType<Directory>()
        .where(
          (Directory d) =>
              !_name(d).startsWith('.') && _file(d, 'meta.json').existsSync(),
        )
        .toList()
      ..sort(
        (Directory a, Directory b) => PyCompat.compare(_name(a), _name(b)),
      );
  }

  /// Initialize and reconcile interrupted work without inferring new permission.
  Future<void> start() {
    if (_stopping) throw const RuntimeError('处理器已停止，请重新打开书库');
    return _initializing ??= _start();
  }

  Future<void> _start() async {
    await reconcile();
    if (_stopping) return;
    _alive = true;
    if (enabled) _timer = Timer.periodic(scanInterval, (_) => _wake());
    _notify();
    if (enabled || _queue.isNotEmpty) _wake();
  }

  /// Reconcile stale UI state only after proving that no live lease owns it.
  /// A completed book's pending quality marker never creates a retry request.
  Future<void> reconcile() async {
    for (final Directory root in _roots()) {
      if (_current == _name(root)) continue;
      try {
        Json state = _json(root, 'status.json');
        if (!<String>{
          'queued',
          'running',
          'finalizing',
          'cancelling',
        }.contains(state['state']))
          continue;
        if (await _probe(root)) continue;
        await _withLease(root, () {
          state = _json(root, 'status.json');
          if (!<String>{
            'queued',
            'running',
            'finalizing',
            'cancelling',
          }.contains(state['state']))
            return;
          final Json meta = _json(root, 'meta.json');
          final String next =
              enabled && _truth(meta['auto']) ? 'queued' : 'paused';
          state.addAll(<String, Object?>{
            'state': next,
            'updated': _clock(),
            'error': null,
          });
          _save(root, 'status.json', state);
          _receipt(root, <String, Object?>{
            'phase': next,
            'reason': 'reconciled',
            'owner_pid': null,
          });
        });
      } on BusyBook {
        continue;
      } on Object catch (error) {
        _lastError = <String, Object?>{
          'book': _name(root),
          'message': _short(error),
          'at': _clock(),
        };
      }
    }
    _notify();
  }

  Future<void> startBook(Directory root) async {
    await start();
    await setAuto(root, true);
  }

  Future<void> resumeBook(Directory root) => startBook(root);

  /// Preserve Python's explicit quality-retry acknowledgement protocol.
  Future<void> setAuto(Directory root, bool value) async {
    _checkRoot(root);
    if (value && _stopping) throw const RuntimeError('处理器已停止，请重新打开书库');
    if (value &&
        _current == _name(root) &&
        (_cancellation?.isCancelled ?? false)) {
      throw const RuntimeError('正在暂停，请等待当前请求结束后再继续');
    }
    final Json meta = _json(root, 'meta.json');
    final Json state = _json(root, 'status.json');
    final Json quality = _object(state['quality']);
    final bool retryQuality =
        value &&
        (_truth(quality['pending']) || quality['state'] == 'pending') &&
        !<String>{
          'running',
          'finalizing',
          'cancelling',
        }.contains(state['state']);
    meta['auto'] = value;
    if (retryQuality) {
      meta['retry_quality'] = true;
    } else if (!value) {
      meta.remove('retry_quality');
    }
    _save(root, 'meta.json', meta);
    final bool queued =
        value &&
        (retryQuality ||
            !<String>{
              'done',
              'running',
              'finalizing',
              'cancelling',
            }.contains(state['state']));
    if (queued) {
      state.addAll(<String, Object?>{
        'state': 'queued',
        'updated': _clock(),
        'error': null,
      });
      _save(root, 'status.json', state);
      _receipt(
        root,
        <String, Object?>{
          'phase': 'queued',
          'retry_quality': retryQuality,
          'owner_pid': null,
        },
        newRequest:
            retryQuality &&
            _json(root, 'work/worker-receipt.json')['phase'] == 'done',
      );
    }
    final String id = _name(root);
    if (!value) {
      _queue.remove(id);
    } else if (_current != id && (queued || !enabled)) {
      _queue.add(id);
    }
    _notify();
    _wake();
  }

  /// Cooperative cancellation waits for the current boundary. A timeout keeps
  /// the slot and run.lock owned until the actual run Future settles.
  Future<void> pauseBook(
    Directory root, {
    Duration timeout = const Duration(seconds: 15),
    bool preserveAuto = false,
  }) async {
    _checkRoot(root);
    // Mark intent before the first await, so a rapid resume cannot flip auto
    // back on while this run is already leaving its cancellation boundary.
    if (_current == _name(root)) _cancellation?.cancel();
    if (!preserveAuto) await setAuto(root, false);
    _queue.remove(_name(root));
    final Completer<void>? active =
        _current == _name(root) ? _currentDone : null;
    if (active != null) {
      _cancellation!.cancel();
      final Json state = _json(root, 'status.json');
      state.addAll(<String, Object?>{
        'state': 'cancelling',
        'updated': _clock(),
      });
      _save(root, 'status.json', state);
      _receipt(root, <String, Object?>{'phase': 'cancelling'});
      _notify();
      try {
        await active.future.timeout(timeout);
      } on TimeoutException {
        return;
      }
    }
    // A PID/status file is not proof of ownership; never stop an external job.
    if (await _probe(root)) throw const BusyBook();
    if (!root.existsSync()) return;
    await _withLease(root, () {
      final Json state = _json(root, 'status.json');
      if (state['state'] != 'done') {
        final String next = preserveAuto ? 'queued' : 'paused';
        state.addAll(<String, Object?>{
          'state': next,
          'updated': _clock(),
          'error': null,
        });
        _save(root, 'status.json', state);
        _receipt(root, <String, Object?>{'phase': next, 'owner_pid': null});
      }
    });
    _notify();
  }

  /// Compatibility spelling for callers migrating server.jobs.cancel.
  Future<void> cancel(
    Directory root, {
    Duration timeout = const Duration(seconds: 15),
    bool preserveAuto = false,
  }) => pauseBook(root, timeout: timeout, preserveAuto: preserveAuto);

  void _wake() {
    if (!_alive || _stopping) return;
    _scanAgain = true;
    _draining ??= _drain().whenComplete(() {
      _draining = null;
      if (_scanAgain && !_stopping) _wake();
    });
  }

  Future<void> _drain() async {
    while (_scanAgain && !_stopping) {
      _scanAgain = false;
      _lastScan = _clock();
      try {
        final List<Directory> roots;
        if (enabled) {
          roots = _roots();
          _queue.clear();
        } else {
          final List<String> ids = _queue.toList()..sort(PyCompat.compare);
          _queue.clear();
          roots = <Directory>[
            for (final String id in ids) Directory('${books.path}/$id'),
          ];
        }
        for (final Directory root in roots) {
          if (_stopping) break;
          try {
            await processBook(root);
          } on Object catch (error) {
            _lastError = <String, Object?>{
              'book': _name(root),
              'message': _short(error),
              'at': _clock(),
            };
            if (root.existsSync()) {
              try {
                final Json state = _json(root, 'status.json');
                state.addAll(<String, Object?>{
                  'state': 'error',
                  'error': '后台处理失败，请查看服务日志',
                  'updated': _clock(),
                });
                _save(root, 'status.json', state);
                _receipt(root, <String, Object?>{
                  'phase': 'error',
                  'owner_pid': null,
                  'error': state['error'],
                });
              } on Object {
                /* Preserve original failure in worker health. */
              }
            }
          }
          _notify();
        }
      } on Object catch (error) {
        _lastError = <String, Object?>{
          'message': _short(error),
          'at': _clock(),
        };
      }
      _notify();
    }
  }

  bool _eligible(Directory root, Json meta, Json state) =>
      _file(root, 'book.json').existsSync() &&
      _truth(meta['auto']) &&
      !(state['state'] == 'done' && !_truth(meta['retry_quality'])) &&
      !(state['state'] == 'error' &&
          _clock() - ((state['updated'] as num?) ?? 0) < 1800);

  /// One serialized attempt, also useful for deterministic offline executors.
  /// This does not grant permission: meta.auto must already be explicit/true.
  Future<void> processBook(Directory root) => _processGate.run(() async {
    if (_stopping || !root.existsSync()) return;
    _checkRoot(root);
    Json meta = _json(root, 'meta.json'), state = _json(root, 'status.json');
    if (!_eligible(root, meta, state) || await _probe(root)) return;
    meta = _json(root, 'meta.json');
    state = _json(root, 'status.json');
    if (_stopping || !_eligible(root, meta, state)) return;
    final bool retryQuality = _truth(meta['retry_quality']);
    final Json? retryBefore =
        retryQuality ? _read(_file(root, 'work/quality-retry.json')) : null;
    final WorkerSettings settings = _settings();
    if (settings.concurrency < 1) throw const ValueError('并发数量必须大于零');
    final RunCancellation cancellation = RunCancellation();
    _current = _name(root);
    _cancellation = cancellation;
    _currentDone = Completer<void>();
    final String? previousJudgeDirectory = environ['JUDGE_LOG_DIR'];
    int code = 1;
    try {
      // A native worker runs several books in one isolate. Give each attempt
      // the cache location and fresh process counters of the original worker.
      environ['JUDGE_LOG_DIR'] = '${root.path}/work/judge';
      jev.resetJevStats();
      final Json prior = _json(root, 'work/worker-receipt.json');
      _receipt(root, <String, Object?>{
        'phase': 'running',
        'owner_pid': pid,
        'attempt': ((prior['attempt'] as int?) ?? 0) + 1,
        'model': settings.model,
        'local_model': settings.localModel,
        'concurrency': settings.concurrency,
        'retry_quality': retryQuality,
      });
      _notify();
      try {
        await _runBook(
          root,
          cancellation: cancellation,
          retryQuality: retryQuality,
          model: settings.model,
          localModel: settings.localModel,
          concurrency: settings.concurrency,
        );
        code = 0;
      } on Cancelled {
        code = 75;
      } on AlreadyRunning {
        code = 75;
      } on Object catch (error) {
        code = 1;
        _lastError = <String, Object?>{
          'book': _name(root),
          'message': _short(error),
          'at': _clock(),
        };
      }
      if (!root.existsSync()) return;
      state = _json(root, 'status.json');
      meta = _json(root, 'meta.json');
      final bool auto = _truth(meta['auto']);
      if (retryQuality && code != 75) {
        final Json retryAfter = _json(root, 'work/quality-retry.json');
        final bool acknowledged =
            <String>{'rebuilding', 'complete'}.contains(retryAfter['state']) &&
            (!_same(retryAfter, retryBefore) ||
                retryBefore?['state'] == 'rebuilding');
        if (acknowledged) {
          meta.remove('retry_quality');
          _save(root, 'meta.json', meta);
        }
      }
      if (auto && _stopping && code != 0) {
        state.addAll(<String, Object?>{
          'state': 'queued',
          'updated': _clock(),
          'error': null,
        });
        _save(root, 'status.json', state);
      } else if (!auto) {
        if (state['state'] != 'done') {
          state.addAll(<String, Object?>{
            'state': 'paused',
            'updated': _clock(),
            'error': null,
          });
          _save(root, 'status.json', state);
        }
      } else if (code != 75 && (code != 0 || state['state'] != 'done')) {
        state.addAll(<String, Object?>{
          'state': 'error',
          'updated': _clock(),
          'error':
              _truth(state['error']) ? state['error'] : '处理进程未完成（退出码 $code）',
        });
        _save(root, 'status.json', state);
        _lastError = <String, Object?>{
          'book': _name(root),
          'message': state['error'],
          'at': _clock(),
        };
      }
      _receipt(root, <String, Object?>{
        'phase': state['state'] ?? (code == 75 ? 'interrupted' : 'error'),
        'exit_code': code,
        'owner_pid': null,
        'finished_at': _clock(),
      });
    } finally {
      if (previousJudgeDirectory == null) {
        environ.remove('JUDGE_LOG_DIR');
      } else {
        environ['JUDGE_LOG_DIR'] = previousJudgeDirectory;
      }
      _current = null;
      _cancellation = null;
      _currentDone!.complete();
      _currentDone = null;
      if (_stopping) _alive = false;
      _notify();
    }
  });

  void _receipt(Directory root, Json fields, {bool newRequest = false}) {
    final Json previous =
        newRequest
            ? <String, Object?>{}
            : _json(root, 'work/worker-receipt.json');
    final double now = _clock();
    _save(root, 'work/worker-receipt.json', <String, Object?>{
      'schema': 1,
      'book': _name(root),
      'request_id': '${(now * 1000000).round()}-$pid-${++_sequence}',
      'requested_at': now,
      'attempt': 0,
      ...previous,
      ...fields,
      'updated': now,
    });
  }

  /// Stop accepting work and preserve automatic intent for an explicit restart.
  /// No model Future is abandoned and no guessed process identifier is killed.
  void stop() {
    _stopping = true;
    _timer?.cancel();
    _timer = null;
    _cancellation?.cancel();
    _queue.clear();
    if (_current == null) _alive = false;
    _notify();
  }

  Future<void> waitIdle() async {
    while (_draining != null || _currentDone != null) {
      await (_draining ?? _currentDone!.future);
    }
  }

  /// For callers that need actual settlement after pauseBook before removal.
  Future<void> waitBookIdle(Directory root) async {
    while (_current == _name(root) && _currentDone != null) {
      await _currentDone!.future;
    }
    if (await _probe(root)) throw const BusyBook();
  }

  /// A timed-out close keeps health.current/lease truthful until run settles.
  Future<void> close({Duration timeout = const Duration(seconds: 15)}) async {
    stop();
    try {
      await waitIdle().timeout(timeout);
    } on TimeoutException {
      /* Still cancelling. */
    }
  }
}
